From cc90b4250bcd84543f77e659049e8fb97bd23171 Mon Sep 17 00:00:00 2001
From: ladehunter <ladehunter@gmail.com>
Date: Wed, 25 Sep 2019 02:44:04 +0000
Subject: [PATCH 109/110] oto8: frameworks: base: Allow system window get focus
 from home to solve the problem

that when Launcher is being focused, startup menu never get keyevent.

Signed-off-by: ladehunter <ladehunter@gmail.com>
Reviewed-by: Liu Xiaoxu <liuxiaoxu@openthos.org>
Reviewed-by: Wang Zhixu <wangzhixu@openthos.org>
---
 .../core/java/com/android/server/wm/WindowManagerService.java  | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/services/core/java/com/android/server/wm/WindowManagerService.java b/services/core/java/com/android/server/wm/WindowManagerService.java
index 4861c47..bd175d8 100644
--- a/services/core/java/com/android/server/wm/WindowManagerService.java
+++ b/services/core/java/com/android/server/wm/WindowManagerService.java
@@ -36,6 +36,7 @@ import static android.os.UserHandle.USER_NULL;
 import static android.view.Display.DEFAULT_DISPLAY;
 import static android.view.Display.INVALID_DISPLAY;
 import static android.view.WindowManager.DOCKED_INVALID;
+import static android.view.WindowManager.LayoutParams.FIRST_SYSTEM_WINDOW;
 import static android.view.WindowManager.LayoutParams.FIRST_APPLICATION_WINDOW;
 import static android.view.WindowManager.LayoutParams.FIRST_SUB_WINDOW;
 import static android.view.WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM;
@@ -5969,12 +5970,13 @@ public class WindowManagerService extends IWindowManager.Stub
 
     // TODO: Move to DisplayContent
     boolean updateFocusedWindowLocked(int mode, boolean updateInputWindows) {
-        WindowState newFocus;
+        WindowState newFocus = mRoot.computeFocusedWindow();
+        boolean isSysmteWindow = newFocus != null
+                && newFocus.getAttrs() != null
+                && newFocus.getAttrs().type >= FIRST_SYSTEM_WINDOW;
         if (mFocusedApp != null && mFocusedApp.getTask() != null
-                                && mFocusedApp.getTask().isHomeTask()) {
+                && !isSysmteWindow && mFocusedApp.getTask().isHomeTask()) {
             newFocus = mFocusedApp.findMainWindow();
-        } else {
-            newFocus = mRoot.computeFocusedWindow();
         }
         if (mCurrentFocus != newFocus) {
             Trace.traceBegin(TRACE_TAG_WINDOW_MANAGER, "wmUpdateFocus");
-- 
2.7.4

